Support for multiple statements in Cassandra cql migration files

agquick opened the pull request:
Currently you cannot include multiple statements in the CQL migration files due to Cassandra only allowing 1 statement per query. This makes creating multiple tables or indices in a single migration difficult. This change splits the CQL migration script by a semi-colon and newline pattern, and iterates them 1 by 1, allowing multiple statements per migration.

agquick commented:
I didn't see that link, that's probably worth adding to the CONTRIBUTING.md file. Responses to your other issues are below:

1. Yes I agree this will break a migration with a string with a semi-colon and newline in it, but this is a pretty uncommon migration. In any event, maybe we could add a flag in the connection url to allow these parsed multi-statement migrations (or some sort of header comment so it only applies to particular migration files?).

2. Yes, this could result in partial changes to the database, but it will just leave the schema in a dirty state, where the user needs to clean it themselves and force the version number. You cannot wrap CREATE TABLE statements in a transaction.

3. A param or header comment for allowing parsed multi-statements will allow old migrations to continue to work as normal. BATCH statements only support INSERT, UPDATE, and DELETE (https://docs.datastax.com/en/cql/3.3/cql/cql_reference/cqlBatch.html).

I agree with your basic concern, but the issue arises from the gocql library (and somewhat Cassandra itself). Apparently, Cassandra can only support a single statement per "session.Query" call. This makes this library very tedious for Cassandra, in that I need to make a separate migration file per individual schema change.

How do you feel about adding a param to the connection url? Or adding a header comment that denotes this migration should be parsed into multiple statements?

(BTW, this issue was brought up before it was forked here (https://github.com/mattes/migrate/issues/338) and here (https://github.com/gocql/gocql/issues/1063)

dhui commented:
Thanks for the context!

I think an option like `x-multi-statement` (disabled by default) for the Cassandra driver, appropriately documented, would work. e.g. warn about simple split on `;` not working w/ embedded semi-colons and atomicity.

Although having headers in the migration file offers more control and granularity, it introduces additional parsing of the migration files which I'd like to avoid to maintain simplicity.

dhui commented on database/cassandra/cassandra.go:
```diff
@@ -155,9 +155,17 @@ func (c *Cassandra) Run(migration io.Reader) error {
-		return database.Error{OrigErr: err, Err: "migration failed", Query: migr}
+
+	// split query by semi-colon
+	queries := strings.Split(query, ";\n")
```
Don't split on `\n` so we don't need to deal with different line endings

agquick commented:
Ok I just pushed support for the `x-multi-statement` flag and updated the README. Also split on ";" instead of ";\n". Ran this on my migrations for my project and it works great.

dhui commented on database/cassandra/README.md:
```diff
@@ -3,6 +3,7 @@
 * Drop command will not work on Cassandra 2.X because it rely on
 system_schema table which comes with 3.X
 * Other commands should work properly but are **not tested**
+* The Cassandra driver (gocql) does not natively support executing multipe statements in a single query. To allow for multiple statements in a single migration, you can use the `x-multi-statement` param. Note that this simply splits the migration into separately-executed statements by a semi-colon ';'. The queries are also not executed in any sort of transaction/batch, meaning you are responsible ...
```
Use `` `;` `` instead of `';'` and provide an explicit warning about semi-colons in strings causing issues.

dhui commented on database/cassandra/cassandra.go:
```diff
@@ -155,9 +157,24 @@ func (c *Cassandra) Run(migration io.Reader) error {
+				return database.Error{OrigErr: err, Err: "migration failed", Query: migr}
+			}
+		}
+	} else {
```
`return` in the `if` block above and remove the `else` to save a level of indentation
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

agquick commented:
Additional requested changes fixed.

dhui commented:
Thanks for your PR!
